### DISCUSSION TITLE
Research into Windows 7 colorization - need answers

### DISCUSSION THREAD

**Discussion by kfh83** [Maplespe/DWMBlurGlass#195] [Q&A] [answered]
So about 3 or 4 days ago me and @aubymori decided to take a look into how we could improve DWMBlurGlass' Aero shader with better colorization. So we found that by default, DWMBG uses just a normal blend mode layer with the tint color and opacity of the user's choice in config.ini/gui. So after researching layer blend modes for a bit we came up with a solution that (almost) seemed to fit the bill: A normal blend mode layer on top, but set at a very low opacity and a multiply layer on the bottom:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/2a25a2e8-f551-4bb1-bfea-458353537836)
As you can see, in this situation it looks fairly accurate but once you try the default Sky color from Windows 7 it quickly falls apart, as it looks too "bright":
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/815f5b07-909a-41a7-8a6b-b8cfaf339600)
(same color, in windows 7:)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/3d3ec14f-ae28-4b6e-a61d-43f5ecdd3b38)
Which we thought was really confusing because how could it look so accurate in another color copied DIRECTLY from the registry, but look so innacurate on another one? Then, after playing with Winaero Tweaker we finally found out there was something other than the 2 layers we already discovered: Blur balance. This is the part that we have no idea what it does or how it works, aswell as the multiply layer on windows 7, which doesnt seem to be controlled by a simple opacity. This control is also relevant in the inactive window state, as it causes the "overexposure" effect that DWMBlurGlass tries to mimic using other effects, but in Windows 7 this seems solely controlled by the multiply layer and whatever is the Blur layer. Here you can see some examples:

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/1b5b61aa-76d9-4857-b4b5-bd90d3584727)
Blur 0, Secondary (multiply) 100, Normal default

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/6bfad6c9-79c0-4871-9309-94a2494aeebb)
Active made to replicate inactive look (Normal set to 40% of original opacity, Multiply set to 60 instead of 43 and blur set to 65 from 49)

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/22005af3-601d-4106-b22f-c0459dacec6c)
Default sky settings

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/56f6782e-1306-425e-8817-c249747ba88d)
Normal layer set to 100 alone

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/552be416-4ed7-49b6-958a-765f5268f0f8)
Multiply set to 100

I dont have more photos, but setting everything to 0 makes the window opaque. Increasing blur only and leaving everything else on 0 seems to make the window go from fully black opaque to fully transparent, but it isn't a simple opacity slider as you can see by increasing the other values with it on 100:

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/076d5490-039c-4ede-b147-b5e50847f4c7)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/937d0d4c-81b2-45e0-b099-3a45d876327f)

So, if anyone (specially the project holders) can look deeper into Windows 7's color effects and give an insight it would be really helpful as we're stuck on this right now, and it would pretty much unlock accurate colorization.

**Comment by kfh83** [comment]
Oh i should probably mention, we are also not sure on how the color "balances" work and what they actually control. Again, doesn't seem like a simple opacity slider.

**Comment by ALTaleX531** [comment]
In fact right now the Aero recipe that DwmBlurGlass is currently using is a bit outdated, our newest recipe is currently located at [AcrylicEverywhere](https://github.com/ALTaleX531/AcrylicEverywhere) and I believe it has a more accurate look than you guys think. You can try compiling it and injecting it into dwm using the injector to see how our newest recipe looks.

**Reply by ALTaleX531** [reply]
<img width="1156" alt="b66cc48530659ae6ac67f6bbd0289755" src="https://github.com/Maplespe/DWMBlurGlass/assets/105535231/e88feaaf-fe21-4fc4-bbdb-2b73498bac5b">
Windows 7
<img width="1156" alt="9c55b91a30bf25f0a8b76c385310cdd3" src="https://github.com/Maplespe/DWMBlurGlass/assets/105535231/7eabfb17-42ff-4863-88aa-c4ed8240323a">
Windows 10 with Aero 10

**Reply by aubymori** [reply]
@ALTaleX531 How does this look on a pure black background?

**Reply by ALTaleX531** [reply]
![image](https://github.com/Maplespe/DWMBlurGlass/assets/105535231/ff533730-c994-4138-b5bc-ad152c91a1ea)
It looks like this.

**Comment by aubymori** [comment]
Here's how my modified DWMBG's shader looks right now. Active is pretty good (though it still needs work), and inactive is way far off.
![image](https://github.com/Maplespe/DWMBlurGlass/assets/44238627/ad8a776c-2dd0-45a7-b160-de2514414b88)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/44238627/cbb0478d-db2c-436e-8389-a6f4e7fc1b77)

**Reply by aubymori** [reply]
This is a build I just made which converts the tint color from the sRGB namespace to the scRGB namespace, which 7 did. That isn't the only thing I changed about the shader though. I made it use multiply blend for the main tint color, and added a translucent normal layer (for the bit of colorization on pure black).

**Comment by kfh83** [comment]
Can you describe the layers used?

**Comment by kfh83** [comment]
> ![image](https://private-user-images.githubusercontent.com/105535231/307546600-ff533730-c994-4138-b5bc-ad152c91a1ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4MzQzMDMsIm5iZiI6MTcwODgzNDAwMywicGF0aCI6Ii8xMDU1MzUyMzEvMzA3NTQ2NjAwLWZmNTMzNzMwLWM5OTQtNDEzOC1iNWJjLWFkMTUyYzkxYTFlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyNVQwNDA2NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZmU5NTg2OWYzNjFmOWU2OThhYmUwOWI0YjIxM2UxNDk4ZjdhYTlhMGQ5YjYwYjRkZDc3NzNhZDFkNDU4YjcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6l0IX3Qyuy3A9wfGyyRjynEU0-wLjE0r15dBUcl1zaM) It looks like this.

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/79777c8f-9ef8-4d3c-8b69-177c1244d396)

Alright, so i retrofitted your implementation from AcrylicEverywhere into our DWMBg version and it seems like this isn't much better than the current default implementation. Looking over it, it seems like you definitely tried to implement some of the ideas from windows 7 (such as the 2 color system and such), but in the end i found it still didn't quite look like it, even though it looks like everything necessary is present. Most notably, when you have a window over a pitch black color, for example, that would barely get colored (due to multiply layer blending), but it seems like it gets just as colored as before with just a normal blend mode layer. As for the inactive, from the screenshots it looks a lot better than what we have right now.

Do you have any clue of what the "Blur" layer might be? Maybe what blend mode it uses?

Oh and also, could you provide the DLL injector you used, as i can't find one that properly injects AcrylicEverywhere, would make testing easier.

**Reply by ALTaleX531** [reply]
![image](https://github.com/Maplespe/DWMBlurGlass/assets/105535231/e71990cb-d6d3-4284-8e1d-3092cdebd69e)
Yes, you reminded me of that, and I have just updated the recipe to make it look like this.
I think the main color layer is using D2D1_BLEND_MODE_MULTIPLY and the glow color layer is using D2D1_BLEND_MODE_COLOR, (please replace it with D2D1_BLEND_MODE_LUMINOSITY in the code because of careless mistakes caused by Microsoft programmers)
As for the blur layer I'm still investigating, but I believe a [color matrix effect](https://microsoft.github.io/Win2D/WinUI2/html/T_Microsoft_Graphics_Canvas_Effects_ColorMatrixEffect.htm) was used rather than an exposure effect.
Then, this is the
[DLL Injector](https://github.com/Maplespe/DWMBlurGlass/files/14395479/DllInjector.zip)) I use. please remember to copy files like dbghelp.dll and symsrv.dll to the `Build` folder or the symbols cannot be loaded.

**Reply by kfh83** [reply]
No, i don't think its using BLEND_MODE_COLOR. If you're talking active, it's most certainly a normal + multiply (and then whatever is Blur). I've read in the past it's something to do with Color Dodge, but in my experience it doesn't look right. BLEND_MODE_COLOR _kinda_ looks right on inactive: 
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/87ae8c55-d6e4-4fde-b748-b7b03a23fb9f)
But, of course, doesn't look quite like real Windows 7 with its 3 layers. In fact, once you start to increase color intensity the normal layer starts to come back:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/c12cdc74-13bf-4976-bcc2-85c517d0496c)

In the sky color, most notably, you can see how when you have an inactive window over a white background there is no coloring:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/135c7a2a-9996-4a84-8309-aace6f7e19c1)
(of course this does not mean every single color is gonna be like that)
This is kinda why it reminds me of BLEND_MODE_COLOR, but again, not quite it. It looks like text and stuff behind it is darker.

Offtopic: here's your inactive formula. It's fairly good:

![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/255d8460-6bca-451d-86ab-3666d8a33506)

... except in dark backgrounds:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/f72f0120-c954-4113-8a03-9ed0b8db450b)
(vs windows 7):
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/dd5529bc-ffad-4f1a-807f-f40442bbc543)

**Reply by ALTaleX531** [reply]
![image](https://github.com/Maplespe/DWMBlurGlass/assets/105535231/5fd0a588-6913-4d9b-8919-1d09fddff8e1)
I remember in my latest recipe that I changed, the inactive coloring doesn't look like this. But yes, my recipe isn't accurate enough, and this is just my initial observation of the Aero, which definitely needs further improvement and more investigation.

**Comment by kfh83** [comment]
@ALTaleX531 Did some slight modifications on your formula and was able to get pretty good results:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/1aa6c494-b6fb-431c-bb98-58cd666fb506)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/1c33188b-037e-4474-a144-be30e33bdaf7)

Windows 7 reference pic:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/85626296/681a33d1-3652-4ff1-837c-c3dd789bf9e8)

It seems like you flipped blurbalance to be negative. Otherwise, the formula stays relatively unchanged except with different values.

**Reply by ALTaleX531** [reply]
Very well then. By the way, I think Aero also includes something for saturation adjustments, since the reduction in `Blur Balance` grays out the backdrop.
![image](https://github.com/Maplespe/DWMBlurGlass/assets/105535231/dfba8b88-b1cc-43a5-a136-451cdc08d171)

**Reply by kfh83** [reply]
Oh that's interesting, i didn't notice that it actually got rid of the saturation completely like this

**Reply by TorutheRedFox** [reply]
fyi doing desaturation in a shader can be done like this

```HLSL
float grayscale = dot(color, float3(0.2125f, 0.7154f, 0.0721f));
float3 desaturated = lerp(grayscale, color, saturate(saturation));
```
where color is the color you're desaturating and saturation is the amount of saturation between 0 (completely grayscale) and 1 (full saturation)

(just to clear things up, saturate() is just the HLSL/Cg name for simply clamping a value between 0 and 1 and I added it as a failsafe as lerp isn't designed to take any values outside of that range)

**Reply by TheInsane101** [reply]
What program are you using for that blank window to test with?

**Reply by kfh83** [reply]
[GlassWnd.zip](https://github.com/Maplespe/DWMBlurGlass/files/14752072/GlassWnd.zip)
By @aubymori

**Reply by TheInsane101** [reply]
Thanks! But is it the same program as ALTaleX531 is using that says "Line"in the titlebar with the toolbar and text below it?

**Comment by j4wassup** [comment]
> @ALTaleX531 Did some slight modifications on your formula and was able to get pretty good results: ![image](https://private-user-images.githubusercontent.com/85626296/307623835-1aa6c494-b6fb-431c-bb98-58cd666fb506.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMjA2OTUsIm5iZiI6MTcwOTEyMDM5NSwicGF0aCI6Ii84NTYyNjI5Ni8zMDc2MjM4MzUtMWFhNmM0OTQtYjZmYi00MzFjLWJiOTgtNThjZDY2NmZiNTA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDExMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxMDFhODI0NzNjMGM2YjYzM2U1MDg2OTFmYTE3ZDEwMTUxMTU5NzIyNDMxZjk3N2YyMTk3OTYyYWI3OGJkNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zHqiRVukOK5d9oOwOQKsAw-zQB5xAYHezvMesGtoQOc) ![image](https://private-user-images.githubusercontent.com/85626296/307623839-1c33188b-037e-4474-a144-be30e33bdaf7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMjA2OTUsIm5iZiI6MTcwOTEyMDM5NSwicGF0aCI6Ii84NTYyNjI5Ni8zMDc2MjM4MzktMWMzMzE4OGItMDM3ZS00NDc0LWExNDQtYmUzMGUzM2JkYWY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDExMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlZTA3ODNlYjE1OTA4YjFkMGM5Mzc1OGIzMjBkZWZjYzM0ZGNiYjkzZWY1NmVhN2M2ZGJiMWIxOWEzNTk4MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jw_MgOFzCpw52H-s5grUzANyvzb1X7cyZqtoKc17LXo)
> 
> Windows 7 reference pic: ![image](https://private-user-images.githubusercontent.com/85626296/307623844-681a33d1-3652-4ff1-837c-c3dd789bf9e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxMjA2OTUsIm5iZiI6MTcwOTEyMDM5NSwicGF0aCI6Ii84NTYyNjI5Ni8zMDc2MjM4NDQtNjgxYTMzZDEtMzY1Mi00ZmYxLTgzN2MtYzNkZDc4OWJmOWU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI4VDExMzk1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThkZWU3NGJkMDFkYmI3NTZlMGFmNTA3OTI5NjI3Yjc3ODg1OTE3Y2IwZTIxZDA1NGRhNTFjZmI0YzA3ZTVhMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LbB-l5ZC3z1h7YuV0EqS0s1cIgSTl7gLj7vg6vSmSjE)
> 
> It seems like you flipped blurbalance to be negative. Otherwise, the formula stays relatively unchanged except with different values.

hey, can i have this one?

**Comment by TorutheRedFox** [comment]
secondary color looks like it might be simply additive?

edit: they all seem to be simply additive, going something like this

(primarycolor * primarybalance) + (desaturatedblur * secondarycolor * secondarybalance) + (blur * blurbalance)

where blur is just the gaussian blurred background with no modifications done to it, not exactly sure how they control the balances based on the opacity slider or inactive state though

**Reply by TorutheRedFox** [reply]
here's me testing the formula in unity (because i have it open lol) and the shader code i used
![image](https://github.com/Maplespe/DWMBlurGlass/assets/15708278/98edbc3d-d924-484a-a753-6c612fa85861)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/15708278/0c3023de-a188-404f-9dd4-b98a6984d94a)
![image](https://github.com/Maplespe/DWMBlurGlass/assets/15708278/cab41f5f-f946-4271-ba4f-4f4179cd512f)

the scaling of the primary balance seems to be non-linear though

**Reply by WackyIdeas** [reply]
>not exactly sure how they control the balances based on the opacity slider or inactive state though

For what it's worth, I ended up making a loose approximation of how the balance values are determined based on the slider position. I simply recorded the values through Winaero Tweaker as I stepped through the intensity slider with the keyboard, which turned out to give 76 steps, and I ended up with an excel sheet that produced this graph:
![image](https://github.com/Maplespe/DWMBlurGlass/assets/60061632/fcc9aa48-8016-449b-b73f-734f0629e334)

The blue line represents the primary color balance, the red line represents the secondary color balance, and the green line represents the blur balance values. I ended up *roughly* approximating this in [Desmos](https://www.desmos.com/calculator/7pz03bat6c) as a bunch of piecewise linear functions within the [0, 1] range. You can see the actual sampled values [here](https://docs.google.com/spreadsheets/d/e/2PACX-1vSlfNz2ZSygjlJhDehM7L_pkEQgMPf5BCpY1BcSPO2JFvg02EC4LeIURORNSGlo_Btt-QrRIPK69V-M/pubhtml). 

Notable things about the behavior:
- Primary, secondary and blur balance values start out at 5%, 5% and 90% respectively
- The primary balance value stays at 5%, the secondary balance linearly climbs while the blur balance linearly drops up until 40% of the slider intensity
- After 40%, the primary balance starts to climb drastically, while the other values begin to drop
- At around 85% of the slider intensity, the secondary color balance drops completely to 0%

As for the inactive state, it seems to be this: 
- Inactive primary color balance = 0.4 * Active color balance
- Secondary color balance remains unchanged
- Inactive blur balance = Active blur balance * 0.4 + 60

**Reply by WackyIdeas** [reply]
Updated the google docs chart with more accurate sampling, and updated the [approximated functions](https://www.desmos.com/calculator/eer8gj14b3) which seem to be much more accurate now, at least for the predefined Win7 colors. Here's a comparison between the actual balance values (left) and the values calculated by the approximations (right). The errors are all off by one errors.
![image](https://github.com/Maplespe/DWMBlurGlass/assets/60061632/fd7f7bba-7e5a-47cf-a2d5-3ff129e9dee1)

**Reply by TheInsane101** [reply]
It would be nice for a small program to be made where you move a slider equivalent to the one in the colour CPL, takes the alpha value and uses it as the x value for all 3 functions simultaneously, and gives the end values of all 3 layers, so you can customise your own DWMBG Aero presets while sticking to 7's formulae. Maybe it could even grab the alpha value from the registry son you can use the native intensity slider?

Update: such a program has been made by CatmanFan

**Reply by TheInsane101** [reply]
I did what you did to see if Windows 7 Milestone 3 build 6801's Aero worked differently, and interestingly, it does! The formula is different. I put the values together in [my own graph,](https://docs.google.com/spreadsheets/d/1MaVqT5w0JeM5ThjH0doJzqAbYeE5pRV_hk1ChnzzF5k/edit?usp=sharing) but I don''t know how to find the equations/functions. How did you figure them out? Thank you for doing so, by the way!

**Reply by kfh83** [reply]
> I did what you did to see if Windows 7 Milestone 3 build 6801's Aero worked differently, and interestingly, it does! The formula is different. I put the values together in [my own graph,](https://docs.google.com/spreadsheets/d/1MaVqT5w0JeM5ThjH0doJzqAbYeE5pRV_hk1ChnzzF5k/edit?usp=sharing) but I don''t know how to find the equations/functions. How did you figure them out? Thank you for doing so, by the way!

6730's aero is even more cryptic as it seems to blend things differently (probably due to a bug being an early implementation of the 3 balance system), maybe one day i'll research into it again but personally i couldn't figure it out.

**Reply by TorutheRedFox** [reply]
it's likely a holdover from Vista, which indeed does it slightly differently

**Reply by kfh83** [reply]
> it's likely a holdover from Vista, which indeed does it slightly differently

Vista's is entirely different as it really only had a control for opacity, 6730 introduced an early version of the 3 way system with all the blending crap but it's entirely different from any build before or after it

**Reply by TorutheRedFox** [reply]
ah

**Comment by kfh83** [comment]
> secondary color looks like it might be simply additive?
> 
> edit: they all seem to be simply additive, going something like this
> 
> (primarycolor * primarybalance) + (desaturatedblur * secondarycolor * secondarybalance) + (blur * blurbalance)
> 
> where blur is just the gaussian blurred background with no modifications done to it, not exactly sure how they control the balances based on the opacity slider or inactive state though

That's the odd thing though because if you crank up the blur blend you'll see it does the "overexposure" effect, idk how to explain since im still lost on it

**Reply by kfh83** [reply]
(altho with just the blur alone it does seem to be just the blurred background alone)

**Reply by TorutheRedFox** [reply]
that's because of the additive blend

if you add two bright images together they'll overexpose, and considering that secondary color is multiplied by the desaturated version of the blur, adding them onto each other creates the overexposure effect if their balances are high

**Comment by kfh83** [comment]
I'll write an explanation soon enough on how Windows 7 colorization works (as far as i know) to maybe bring closure to this. Thanks a lot to @WackyIdeas, @TorutheRedFox for major leads, @ALTaleX531 for helping me with shader work on DWMBlurGlass and to anyone who has contributed with information on this discussion.

**Comment by ALTaleX531** [comment, marked as answer]
Well, I spent all morning today extracting and disassembling the Windows 7 blur shader bytecode. 
I finally got the following recipe, and I'll pick a time to implement it.

```mermaid

flowchart LR;
    backdrop-->|gaussian_blur|blurred_backdrop;
    blurred_backdrop-->|greyscale|greyscaled_backdrop;
    greyscaled_backdrop-->|multiply|blurred_backdrop_multiplied_by_balanced_afterglow;
    blurred_backdrop-->|multiply|blurred_backdrop_multiplied_by_blur_balance;
    blurred_backdrop_multiplied_by_balanced_afterglow-->|add|composited_blurred_backdrop
    blurred_backdrop_multiplied_by_blur_balance-->|add|composited_blurred_backdrop
    balanced_color-->|add|output;
    composited_blurred_backdrop-->|add|output;
```

**Reply by kfh83** [reply]
Again, what i said in the OpenGlass discussion, it seems we have been pretty much spot on for the past few months.

**Reply by ALTaleX531** [reply]
Yes, among us, @TorutheRedFox's [answer](https://github.com/Maplespe/DWMBlurGlass/discussions/195#discussioncomment-8752117) is the most accurate, almost exactly like the original Windows 7 shader, except for some unnecessary calculations and incorrect values.

I'll probably look at how active/inactive colorization is handled in udwm next, it seems to use the theme's data for calculations.

**Reply by TheInsane101** [reply]
Removing the unnecessary calculations should help with performance, so adding it will be great!

**Reply by kfh83** [reply]
> Yes, among us, @TorutheRedFox's [answer](https://github.com/Maplespe/DWMBlurGlass/discussions/195#discussioncomment-8752117) is the most accurate, almost exactly like the original Windows 7 shader, except for some unnecessary calculations and incorrect values.
> 
> I'll probably look at how active/inactive colorization is handled in udwm next, it seems to use the theme's data for calculations.

I don't think (atleast in the openglass legacy implementation me/wiktor came up with) there are any differences to what you just came up with, nor do i recall any unnecessary calculations. You're also probably going to come up with the same thing currently implemented for active/inactive logic, it's just a switch of values.

![image](https://github.com/user-attachments/assets/54cf6ea5-7750-44a0-abff-1fb654ab56b1)
This is (off the top of my memory) what OpenGlass Legacy does currently.

**Reply by ALTaleX531** [reply]
I didn't say anything about unnecessary calculations or incorrect values in your/wiktor's implementation, I'll continue to look into colorization simply out of curiosity and for deeper insight into opaque and animation cases.
